storage-node-v2/src/services/sync/storageObligations.ts

```diff
@@ -0,0 +1,222 @@
+import { QueryNodeApi } from '../queryNode/api'
+import logger from '../logger'
+import {
+  StorageBagDetailsFragment,
+  StorageBucketDetailsFragment,
+  DataObjectDetailsFragment,
+} from '../queryNode/generated/queries'
+
+/**
+ * Defines storage provider data obligations.
+ */
+export type DataObligations = {
+  /**
+   * All storage buckets in the system.
+   */
+  storageBuckets: StorageBucket[]
+
+  /**
+   * Assigned bags for the storage provider.
+   */
+  bags: Bag[]
+
+  /**
+   * Assigned data objects for the storage provider.
+   */
+  dataObjects: DataObject[]
+}
+
+/**
+ * Storage bucket abstraction.
+ */
+type StorageBucket = {
+  /**
+   * Storage bucket ID
+   */
+  id: string
+
+  /**
+   * Storage operator URL
+   */
+  operatorUrl: string
+
+  /**
+   * Storage working group ID.
+   */
+  workerId: number
+}
+
+/**
+ * Storage bag abstracton.
+ */
+type Bag = {
+  /**
+   * Storage bag ID
+   */
+  id: string
+
+  /**
+   * Assigned storage bucket IDs.
+   */
+  buckets: string[]
+}
+
+/**
+ * Data object abstraction.
+ */
+type DataObject = {
+  /**
+   * Data object ID
+   */
+  id: string
+
+  /**
+   * Assigned bag ID
+   */
+  bagId: string
+}
+
+/**
+ * Get storage provider obligations like (assigned data objects) from the
+ * runtime (Query Node).
+ *
+ * @param queryNodeUrl - Query Node URL
+ * @param workerId - worker ID
+ * @returns promise for the DataObligations
+ */
+export async function getStorageObligationsFromRuntime(
+  queryNodeUrl: string,
+  workerId: number
+): Promise<DataObligations> {
+  const api = new QueryNodeApi(queryNodeUrl)
+
+  const allBuckets = await getAllBuckets(api)
```

Comment:
On the other hand, this will remove local non-active buckets or buckets without metadata that will lead to deletion of the local data. I'm not sure that either of these conditions should be implemented. Client-side filtering does the trick.